src/Features/Core/Portable/Formatting/AbstractFormattingDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Formatting
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    internal class AbstractFormattingDiagnosticAnalyzer
+        : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        public AbstractFormattingDiagnosticAnalyzer()
+            : base(
+                IDEDiagnosticIds.FormattingDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Formatting_document), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Formatting_document), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SyntaxAnalysis;
+
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxTreeAction(AnalyzeSyntaxTree);
+
+        private void AnalyzeSyntaxTree(SyntaxTreeAnalysisContext context)
+        {
+            if (!(context.Options is WorkspaceAnalyzerOptions workspaceAnalyzerOptions))
+            {
+                return;
+            }
+
+            var options = context.Options.GetDocumentOptionSetAsync(context.Tree, context.CancellationToken).GetAwaiter().GetResult();
+            if (options == null)
+            {
+                return;
+            }
+
+            var workspace = workspaceAnalyzerOptions.Services.Workspace;
+            var formattingChanges = Formatter.GetFormattedTextChanges(context.Tree.GetRoot(context.CancellationToken), workspace, options, context.CancellationToken);
+            foreach (var change in formattingChanges)
+            {
+                var span = change.Span;
+                if (span.IsEmpty)
+                {
+                    // Need to add characters
+                    span = new TextSpan(change.Span.Start, 1);
+                }
+
+                var location = Location.Create(context.Tree, change.Span);
+                context.ReportDiagnostic(DiagnosticHelper.Create(
```

Comment:
The initial implementation is aggressive. I was _certain_ I would hate it, but when I ran it across a variety of projects it was the correct and acceptable behavior far more often than not. Based on the experimental results, I removed selective configurability from the requirements for the initial experimental-use-only MVP.

Reply:
i think part of hte aggressiveness concern is that this can lead to something like an error *per line*, even when most of the actual *visible* C# code is actually formatted just fine.

That's why this seems problematic to me.  The code you can see may look totally ok.  But it's code that you don't see that is a concern.  Note, this is not because it's whitespace that we're touching.  For example, if you have:

```for  (var a = 1;a >10;a++)```

I can see many of the whitespace problems there, and i do want htem fixed up.

However, i can't actually see a problem at the end of the line (which is likely why so much code actually has whitespace issues there).

Reply:
Another data point: when we have an issue with 'usings' we don't report a sorting problem on each using that is non-sorted.  we just report on the first, and we allow you to fix it.  This helps give the user the relevant info, without giving them a deluge of problems.  VSCode does *not* do this, and i hate it (since i can get like 10+ messages about a section of imports being a problem).

I think we do have to think about how we can give hte user the relevant info, without overwhelming with diagnostics.

Reply:
are you saying that it is good that it is reporting trailing whitespace?

> However, i can't actually see a problem at the end of the line (which is likely why so much code actually has whitespace issues there

this kind of thing is what analyzer trying to help...

Reply:
> issues I can see and know right away. they don't need any analyzer help since I can figure out right away.

But we do.  As we can see from runnign this ourselves, people have checked in things like ```if(```.  This is a problem because the actual tokens on the line don't match our formatting expectation.

The problem here is that we are assuming that 'formatting' for people means 'ensuring *all* whitespace is correct'.  i would counter that by saying:  it's an acceptable interpretation to think that formatting means 'my actual code is indented and spaced out properly'.  

In other words, i do not think about formatting in terms of whitespace itself.  I think about it in terms of my *actual code* (i.e. *non-whitespace*), and how that is laid out.

Both are reasonable interpretations.  However, the former interpretation views some code as incorrect, even if someone might feel it was totally fine.  I would like it to be actually discussed and weighed if we should be enforcing the former view rather than the more permissive latter view. (Or, if we should have an option to allow the user to decide which view they care about).

Reply:
As another data point: 

>  they don't need any analyzer help since I can figure out right away.

This is clearly untrue.  We report indentation/alignment/spacing issues *all the time* in PRs.  However, people normally don't bother reporting "trailing whitespace" issues in PRs.  Because, by default, tooling (including github) doesn't even show it.

By enforcing a trailing-whitespace rule, we are saying: even though so many tools don't even show this by default, you must now fix this.  This is effectively enforcing the *former* view on source-code formatting, when (by default) most people look at code through the *latter* view.

Note: the latter view is totally acceptable.  It can even be something that is turned on by default.  However, i don't think it's a universal view, and i think users shoudl be able to say that it is not their view.

Reply:
the way I see it is that there are so many trailing whitespace issues compared to other whitespace issues. meaning other whitespace issues are much easier to spot compared to trailing whitespace. since tool should help where it hurts most, detecting trailing whitespace issue seems most useful one to have.

sure there will be other whitespace cases that are not formatted correctly since human makes mistake even for obvious things time to time. but I believe the more important thing is what errors/issues happen more frequently.

...

like I said, C# IDE has been removing trailing whitespace since Whidbey or before. I am not sure why github wins over years of experience VS IDE has been provided.

all VS tool shows trailing space changes by the way.

Reply:
so, I believe it is fine that we first address the people who are used to VS C# IDE and then address people you are describing? such as  someone who doesn't care using tab/space, or long whitespaces in blank lines and etc once we have feedbacks from v1?

Reply:
> so, I believe it is fine that we first address the people who are used to VS C# IDE

By default, the C# ide makes trailing whitespace invisible.  I do not think that we should enforce the view that invisible code is code you must fix if you want this analyzer.  I think it's ok to do that *by default*.  But i think users should have an option to say: i do not care about invisible code, please do not pester me.

--

The point of me talking aabout *other* tools was simply to reinforce how common the view is that trailing whitespace is not something traditionally brought up to the level that someone should be cognizant of it.  VS is consistent with other tools in that manner.  I think roslyn should have a way to say that such a policy also makes sense for it as well.  I do not think roslyn should force people to now have squiggles on invisible code just because htey wanted *visible* code to be formatted properly.

Reply:
Yeah, i've looked at nearly a dozen of tools/IDEs now.  As far as i can tell, the default policy is that trailing whitespace is invisible and does not rise to the level of something that users traditionally need to interact with.  I do not think we should then be reporting diagnostics on this *unless* users have a way to say: that's nice and all, but i don't care about these issues.  I care about actual code *i can see* that violates my formatting options.

Reply:
it is not just trailer whitespace. it is mix usage of tab/spaces. it is unnecessary whitespaces in blank lines and etc.

...

I am saying v2 since that require more work for analyzer to determine kind of changes it needs to do. and we should do that only if we get enough feedbacks on it rather than speculating. 

this now becomes design questino so, tagging @jinujoseph  and @kuhlenh

Reply:
> it is unnecessary whitespaces in blank lines and etc.

Which i think is simply trailing whitespace...

> I am saying v2 since that require more work for analyzer to determine kind of changes it needs to do.  

It would literally be a dozen lines of code or so.  We shoudl really not be averse to spending a little effort to make good experiences.  Similar to conversations we've had in the past *more* effort has been spent arguing, than just going and building a better experience.  If we can put in the energy to argue this must, we can just go and handle these things.
> and we should do that only if we get enough feedbacks on it rather than speculating.

I think we can effectively look at the ecosystem out there and make design decisions without continually needing users to have to pester us about these things.  

this now becomes design questino so, tagging @jinujoseph and @kuhlenh

Indeed.  I've been clear that this is a design decision.  And i've laid out my reasons for why i think this is important.  We have ample evidence across many tools that this is the default way that code is presented across the ecosystem.  We should be cognizant of this.

Note: this should also tie into something like trim_trailing_whitespace=true/safe.  If someone has proactively chosen such a setting, then we should of course respect it.  But if someone has made no indication that invisible code is relevant, then we should be wary about introducing a feature that now forces them to care *without* giving a way to relax that behavior.